### PR TITLE
fix flaky 03172_error_log_table_not_empty

### PR DESCRIPTION
**Pull request by CheSema**
Here is the details
https://github.com/ClickHouse/ClickHouse/pull/65604#discussion_r1665690244

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by pamarcos** [comment]
I don't think this is fixing the test, to be honest. I don't understand neither of the 2 changes.

Why do you need to explicitly flush first thing if the test is already making sure logs are flushed before reading from `system.errors_log`?

Regarding removing the `sleep(2)`, that should do the test way more flaky. Take into account that you can flush the table, but you need to give it time for error logs to be collected minimum at the interval time specified in the settings so that something at all is flushed. It's exactly the same as for [00990_metric_log_table_not_empty](https://github.com/ClickHouse/ClickHouse/blob/19816401c4b1a3a74c3fcfd0ed6dfde4f405fcf9/tests/queries/0_stateless/00990_metric_log_table_not_empty.sql)

**Comment by CheSema** [comment]
> I don't think this is fixing the test, to be honest. I don't understand neither of the 2 changes.
> 
> Why do you need to explicitly flush first thing if the test is already making sure logs are flushed before reading from `system.errors_log`?

That is the line when the read from `system.error_log` is made before writing to it.
https://github.com/ClickHouse/ClickHouse/blob/1e5bc5bc8c7e65908e7127e06e1f5008f6fbfa20/tests/queries/0_stateless/03172_error_log_table_not_empty.sh#L11

> Regarding removing the `sleep(2)`, that should do the test way more flaky. Take into account that you can flush the table, but you need to give it time for error logs to be collected minimum at the interval time specified in the settings so that something at all is flushed. It's exactly the same as for [00990_metric_log_table_not_empty](https://github.com/ClickHouse/ClickHouse/blob/19816401c4b1a3a74c3fcfd0ed6dfde4f405fcf9/tests/queries/0_stateless/00990_metric_log_table_not_empty.sql)

`SYSTEM FLUSH LOGS` flushes the data and waits until it done. No need to sleep.

**Comment by pamarcos** [comment]
> `SYSTEM FLUSH LOGS` flushes the data and waits until it done. No need to sleep.

It's not that the test only needs to wait for the data to be flushed. It needs to wait so that the data is collected as well. If you don't wait, no data will be collected. Forcing a flush doesn't collect the data, since it always relies on periodic intervals to collect data.

**Comment by CheSema** [comment]
> SYSTEM FLUSH LOGS

Please define what you mean by collecting data?

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 1e5bc5bc8c7e65908e7127e06e1f5008f6fbfa20 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66093/1e5bc5bc8c7e65908e7127e06e1f5008f6fbfa20/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Integration tests</td><td>The integration tests report. In parenthesis the package type is given, and in square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/integration_tests__aarch64__[1_6].html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/performance_comparison__release__[2_4]/report.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/975e8ee14a16c0ee7c8c558dd40f0fef1ef10a4e/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by pamarcos** [comment]
> Please define what you mean by collecting data?

Collect the errors](https://github.com/ClickHouse/ClickHouse/blob/196df125b1e9fa9a5292cdb27ab479c6aeef1be6/src/Interpreters/ErrorLog.cpp#L88-L121). This method is called by a background thread periodically.

File: tests/queries/0_stateless/03172_error_log_table_not_empty.sh [hunk 1]
```diff
@@ -4,17 +4,19 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
...
+# system.error_log is created lazy, flush logs query makes it sure that the table is created.
+$CLICKHOUSE_CLIENT -q "SYSTEM FLUSH LOGS;"
+
 # Get the previous number of errors for 111, 222 and 333
 errors_111=$($CLICKHOUSE_CLIENT -q "SELECT sum(value) FROM system.error_log WHERE code = 111")
 errors_222=$($CLICKHOUSE_CLIENT -q "SELECT sum(value) FROM system.error_log WHERE code = 222")
 errors_333=$($CLICKHOUSE_CLIENT -q "SELECT sum(value) FROM system.error_log WHERE code = 333")
 
-# Throw three random errors: 111, 222 and 333 and wait for more than collect_interval_milliseconds to ensure system.error_log is flushed
+# Throw three random errors: 111, 222 and 333 and call flush logs to ensure system.error_log is flushed
```
Comment (pamarcos) [inline tests/queries/0_stateless/03172_error_log_table_not_empty.sh:15]:
```suggestion
# Throw three random errors: 111, 222 and 333, wait for more than collect_time_interval_milliseconds to ensure the errors are collected and then flush logs to ensure system.error_log is flushed
```

**Comment by CheSema** [comment]
> > Please define what you mean by collecting data?
> 
> [Collect the errors](https://github.com/ClickHouse/ClickHouse/blob/196df125b1e9fa9a5292cdb27ab479c6aeef1be6/src/Interpreters/ErrorLog.cpp#L88-L121). This method is called by a background thread periodically.

Oh, that is a problem.
We have an invariant that `SYSTEM FLUSH LOGS` does its jobs synchronously. It happened that it is wrong for `system.error_log`.
Usually all events for `system.[some_log]` are accumulated in the buffer, which is written by `SYSTEM FLUSH LOGS`.
Your are pointing that there is no such accumulating process.

File: tests/queries/0_stateless/03172_error_log_table_not_empty.sh [hunk 2]
```diff
@@ -4,17 +4,19 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
...
 $CLICKHOUSE_CLIENT -mn -q "
 SELECT throwIf(true, 'error_log', toInt16(111)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 111 }
 SELECT throwIf(true, 'error_log', toInt16(222)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 222 }
 SELECT throwIf(true, 'error_log', toInt16(333)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 333 }
-SELECT sleep(2) format NULL;
```
Comment (pamarcos) [inline tests/queries/0_stateless/03172_error_log_table_not_empty.sh:17]: We need this. Without it, no errors are collected and tests fail close to 100% of the time:

```bash
+++ /home/ubuntu/ClickHouse/ClickHouse/tests/queries/0_stateless/03172_error_log_table_not_empty.stdout 2024-07-04 16:29:37.956423656 +0000
@@ -1,6 +1,6 @@
-1
-1
-1
-1
-1
-1
+0
+0
+0
+0
+0
+0
```

File: tests/queries/0_stateless/03172_error_log_table_not_empty.sh [hunk 3]
```diff
@@ -30,12 +32,11 @@ $CLICKHOUSE_CLIENT -mn -q "
...
```
Comment (pamarcos) [inline tests/queries/0_stateless/03172_error_log_table_not_empty.sh:33]: Same for this one. We actually need it.

**Comment by pamarcos** [comment]
> Oh, that is a problem. We have an invariant that `SYSTEM FLUSH LOGS` does its jobs synchronously. It happened that it is wrong for `system.error_log`. Usually all events for `system.[some_log]` are accumulated in the buffer, which is written by `SYSTEM FLUSH LOGS`. Your are pointing that there is no such accumulating process.

[system.error_log](https://github.com/ClickHouse/ClickHouse/issues/64501) was developed based on [system.metric_log](https://github.com/ClickHouse/ClickHouse/blob/196df125b1e9fa9a5292cdb27ab479c6aeef1be6/src/Interpreters/MetricLog.cpp#L59-L85) which does the same thing. Both are periodic metrics that aggregate the metrics for the defined time window interval. They're different by nature to most other system logs. It doesn't make sense to collect and flush data during a `SYSTEM FLUSH LOGS` if the periodic time interval has not elapsed yet, IMHO.

**Comment by CheSema** [comment]
I have restored timeouts.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by CheSema** [comment]
Logical error:  isUniqTypes()
https://s3.amazonaws.com/clickhouse-test-reports/66093/78a2139f2a43752196a029995b6965ada359c954/stress_test__tsan_.html

**Comment by CheSema** [comment]
> > Oh, that is a problem. We have an invariant that `SYSTEM FLUSH LOGS` does its jobs synchronously. It happened that it is wrong for `system.error_log`. Usually all events for `system.[some_log]` are accumulated in the buffer, which is written by `SYSTEM FLUSH LOGS`. Your are pointing that there is no such accumulating process.
> 
> [system.error_log](https://github.com/ClickHouse/ClickHouse/issues/64501) was developed based on [system.metric_log](https://github.com/ClickHouse/ClickHouse/blob/196df125b1e9fa9a5292cdb27ab479c6aeef1be6/src/Interpreters/MetricLog.cpp#L59-L85) which does the same thing. Both are periodic collectors that aggregate the metrics for the defined time window interval. They're different by nature to most other system logs. It doesn't make sense to collect and flush data during a `SYSTEM FLUSH LOGS` if the periodic time interval has not elapsed yet, IMHO.

It took me a while to understand, you are right here.

**Pull request merged**